xtask/src/ast_src.rs

```diff
@@ -533,70 +1609,359 @@ pub(crate) const AST_SRC: AstSrc = AstSrc {
+        /// ❰ use { blin::blen, blah::* };
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/items/use-declarations.html)
         struct UseItem: AttrsOwner, VisibilityOwner {
             T![use],
             UseTree,
         }
+
+        // TODO: verify example correctness
+        /// Use tree.
+        ///
+        /// ```
+        /// pub use ❰ foo::❰ * ❱ ❱;
+        /// use ❰ bar as baz ❱;
+        /// use ❰ bruh::bruuh::{ ❰ self ❱, ❰ blin ❱ } ❱;
+        /// use ❰ { ❰ blin::blen ❱ } ❱ // TODO: clarify if top-level curlies are `UseTree`
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/items/use-declarations.html)
         struct UseTree {
             Path, T![*], UseTreeList, Alias
         }
+
+        /// Item alias.
+        /// Note: this is not the type alias.
+        ///
+        /// ```
+        /// use foo ❰ as bar ❱;
+        /// use baz::{bruh ❰ as _ ❱};
+        /// extern crate bruuh ❰ as blin ❱;
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/items/use-declarations.html)
         struct Alias: NameOwner { T![as] }
+
+        /// Sublist of use trees.
+        ///
+        /// ```
+        /// use bruh::bruuh::❰ { ❰ self ❱, ❰ blin ❱ } ❱;
+        /// use ❰ { blin::blen::❰ {} ❱ } ❱
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/items/use-declarations.html)
         struct UseTreeList { T!['{'], use_trees: [UseTree], T!['}'] }
+
+        /// Extern crate item.
+        ///
+        /// ```
+        /// ❰ #[attr] pub extern crate foo; ❱
+        /// ❰ extern crate self as bar; ❱
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/items/extern-crates.html)
         struct ExternCrateItem: AttrsOwner, VisibilityOwner {
             T![extern], T![crate], NameRef, Alias,
         }
+
+        /// Call site arguments list.
+        ///
+        /// ```
+        /// foo::<T, U>❰ (42, true) ❱;
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/expressions/call-expr.html)
         struct ArgList {
             T!['('],
             args: [Expr],
             T![')']
         }
+
+        // TODO: correct the example
+        /// Path to a symbol. Includes single identifier names and elaborate paths with
+        /// generic parameters.
+        ///
+        /// ```
+        /// (0..10).❰ collect ❰ ::<Vec<_>> ❱ ❱();
+        /// ❰ Vec ❰ ::<u8> ❰ ::with_capacity ❱ ❱ ❱(1024);
+        /// ❰ <Foo as Bar> ❰ ::baz ❱ ❱();
+        /// ❰ <bruh> ❰ ::bruuh ❱ ❱();
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/paths.html)
         struct Path {
             segment: PathSegment,
             qualifier: Path,
         }
+
+        // TODO: verify the example
+        // TODO: what RetType is doing here? is this for Fn() -> T syntax?
```

Comment:
yes, as is ParamList

Reply:
@flodiebold as I see the colon is not included in `PATH_SEGMENT`
e.g. this code
```rust
pre_commit::run_hook();
```
produces
```
                CALL_EXPR@818..840
                  PATH_EXPR@818..838
                    PATH@818..838
                      PATH@818..828
                        PATH_SEGMENT@818..828
                          NAME_REF@818..828
                            IDENT@818..828 "pre_commit"
                      COLON2@828..830 "::"
                      PATH_SEGMENT@830..838
                        NAME_REF@830..838
                          IDENT@830..838 "run_hook"
                  ARG_LIST@838..840
                    L_PAREN@838..839 "("
                    R_PAREN@839..840 ")"
```
So is this a mistake that there is a `::` in `PathSegment` ast node delcaration and I should not include this token in my example? @matklad

Reply:
Ah, so the absolute path does produce a `::` in `PathSegment`:
```rust
::foo::bar
```
```
          PATH@16..26
            PATH@16..21
              PATH_SEGMENT@16..21
                COLON2@16..18 "::"
                NAME_REF@18..21
                  IDENT@18..21 "foo"
            COLON2@21..23 "::"
            PATH_SEGMENT@23..26
              NAME_REF@23..26
                IDENT@23..26 "bar"
```

Reply:
@matklad is this only once special case when `::` token is included in `PathSegment`:

Reply:
See my changes in a19b164661675be6a03cacfbc652cbd5f06492a0

Reply:
Yup, `::` is for absolute segments. Path grammar has been a pain in the back historically for me :(